### ISSUE TITLE
Connecting Kura to Hono

### ISSUE THREAD

**Issue by mohabh88** [eclipse/kura#2177]
Hello there, 

I am trying to connect my Kura gateway installed on Raspberry Pi 2 B+ to Hono. However, I could not do that. Is there any detailed documentation on how that is done? That would be very helpful. 

For instance, I tried to have `minikube` and `kubectl` installed on the same Raspberry Pi, but unfortunately, every time I start them, I get the below errors:

```
$ minikube start
bash: /usr/local/bin/minikube: cannot execute binary file: Exec format error
``` 

```
$ kubectl version
Illegal instruction
```

I see in this [documentation](https://www.eclipse.org/hono/user-guide/kura-adapter/) that it is possible to connect Kura to Hono via MQTT broker directly. 

So, is it possible to have Hono on Raspberry Pi? Otherwise, what is the best way to have them (i.e., Kura and Hono) connected together? I am using [Raspbian Jessie 2015-11-21](http://downloads.raspberrypi.org/raspbian/images/raspbian-2015-11-24/) as the OS for the Raspberry Pi to have Kura running on top of it following this Eclipse IoT [tutorial](https://iot.eclipse.org/java/tutorial/). 

Basically, I have Hono installed and running on my local machine via Minikube (Kubernetes) as well as Minishift (OpenShift) and I want to connect them to Kura (running on Raspberry Pi) to form a cluster.

I have searched for this information but nowhere mentions Kura and Hono altogether in details. Thank you in advance and appreciate if you help me.

I look forward to hearing from you,
Mohab

**Comment by mohabh88** [comment]
I hope that the below information would help further. Here is what I have:

- [ ] Hono is up and running on a local machine (minikube instance).

- [ ] Kura is up and running on a Raspberry Pi (RPI). 

- [ ] Raspberry Pi IP address (connected to Ethernet): `132.207.28.202`.
Raspberry Pi IP address (connected to wlan0): `132.207.243.109`.
(N.B.: Raspberry Pi is plugged into Ethernet cable and also is equipped with a wifi adapter)

- [ ] Local machine IP address: `132.207.170.30` (where Hono is deployed).
(N.B.: Local machine is plugged into another Ethernet cable).
* minikube IP address: `192.168.99.100`.

- [ ] hono-adapter-kura: cluster IP address: `10.109.198.68`.
hono-adapter-kura: internal endpoints: `hono-adapter-kura.hono:`(`1883`, `31884`, `88831`, `30884`).
* Network: node minikube: `172.17.0.18`.

- [ ] MqttDataTransport (Kura on Raspberry Pi): broker-url: `mqtt://iot.eclipse.org:1883/` (Status: Connected), but when trying to feed the broker-url with any of Hono above addresses! Kura's status changes to Disconnected. May be because they are not on the same network or so?!

Even I tried to plug the RPI into the local machine directly; the IP address got changed and tried to put the IP address of the minikube instance (e.g., `192.168.99.100`) in the broker-url for Kura, but that didn't work as well.

I would like to know how to connect Kura (on RPI) to Hono (minikube and-or minishift instances -- on a local machine) so that the cluster is formed. I'd be grateful and thankful if you shed the light on such integration and if you assisted me. 

Awaiting for your reply and thank you so much for your consideration in advance. Mohab

**Comment by MMaiero** [comment]
Hi, 
I believe the issues are all related to the networking setup. Are you sure that the Hono services are effectively exposed outside to the network. Do you have proper routing rules?

Maybe you can try a simpler setup to test that the Kura-Hono integration works and start making things more complicated step by step.

Best regards,
Matteo

**Comment by mohabh88** [comment]
Hi @MMaiero, thank you so much for replying back. 

Could you please further explain you point? How would I expose out Hono services to the network? Possible if you point me to that (documentations, tutorials, etc.)? That would be greatly appreciated. 

I have very small setup, just the Kura on the Raspberry Pi and Hono is on my local machine and I would like to have them integrated. 

Here you are a screenshot of my Raspberry Pi pinging my local machine... the ping is done successfully. So I might assume there is no issue pertaining to the network setup, right?

![2018-08-08-093545_1280x1024_scrot](https://user-images.githubusercontent.com/9646161/43841099-5a8b1c74-9af0-11e8-9f0f-c7e4c9beac4a.png)

**Comment by MMaiero** [comment]
Hi,
if you can perform a telnet from the raspberry to the computer that is hosting Hono on port 1883 and it is working, it means that the ports are correctly exposed and that Kura should be able to connect there.

**Comment by mohabh88** [comment]
Thank you for your prompt reply. 

No, unfortunately, could not perform a telnet to my local machine (where minikube node is hosted) on the specified port "1883". Also tried to telnet minikube node itself, but still. 

![2018-08-08-103402_1280x1024_scrot](https://user-images.githubusercontent.com/9646161/43844489-d9545280-9af7-11e8-9c24-eb030ebbfff1.png)

Furthermore, tried to re-install talent to make sure that it is working properly, but still I get the same above results. Any other suggestions please?

**Comment by MMaiero** [comment]
As said, I believe you should check the networking configuration in your kubernetes instance in order to effectively expose the hosted hono service to the outside world.

Maybe refer to the Hono team to have more details, if your setup is one of the ones that support.

Best regards,
Matteo

**Comment by MMaiero** [comment]
Hi,
can this be closed now?

Best regards,
Matteo

**Comment by mohabh88** [comment]
Thank you, Matteo, for the heads-up.

Still the issue is not fully resolved yet. But before we have this issue closed, have you ever tried to use [Eclipse Hono Sandbox](https://www.eclipse.org/hono/sandbox/) and connect it to Kura?

There is an issue while connecting Hono to Kura via port-forwarding and maybe one of the alternatives and-or workarounds is to involve Hono Sandbox. I used it with docker swarms as explained in the documentation, but still the connection part is not complete at mine yet! Possible if you walked through this?

Mohab

**Comment by MMaiero** [comment]
Hi,
I've done tests with Kura and the Hono swarm setup for development purposes, but not using the Hono connection for Kura. And it seemed to be working fine.

Best regards,
Matteo

**Comment by mohabh88** [comment]
Thank you so much Matteo, for the prompt reply. Greatly appreciated. 

For me, I am not able to connect to the Sandbox. Even I am using MQTTFx and MQTTLens (for testing and see whether able to connect to the server), and no able is able to connect using the broker-url, port-number and the credentials.

What I am using in MQTTFx:

Broker-address: hono.eclipse.org
Broker-port: 1884
Client ID: 4711

-- User Credentials tab:
Username: sensor1@DEFAULT_TENANT
Password (I am trying with different ones, but no luck) ... what I am trying is:
1) `hono-secret `
2) `verysecret`
3) `jZxT1LMIku7OjiOWEJqRiWrUxSwDy444S8xz9GLKvz9hoaOsxCjhUGtCJD4+cowTjzzvz2piN424BZLsFoWkSA==`
4) `hashed-password`

and the logs says (`Broker Unavailable - Please verify your Settings (e.g. Broker Address, Broker Port & Client ID) and the user credentials!`)

```
2018-08-27 10:46:18,617  INFO --- ScriptsController              : Clear console.
2018-08-27 10:46:18,618 ERROR --- BrokerConnectService           : Broker unavailable
2018-08-27 10:47:04,358  INFO --- BrokerConnectorController      : onConnect
2018-08-27 10:47:04,360  INFO --- ScriptsController              : Clear console.
2018-08-27 10:47:04,361  INFO --- MqttFX ClientModel             : MqttClient with ID 4711 assigned.
2018-08-27 10:47:04,485 ERROR --- MqttFX ClientModel             : Error when connecting
org.eclipse.paho.client.mqttv3.MqttException: Broker unavailable
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:31) ~[org.eclipse.paho.client.mqttv3-1.2.0.jar:?]
	at org.eclipse.paho.client.mqttv3.internal.ClientState.notifyReceivedAck(ClientState.java:988) ~[org.eclipse.paho.client.mqttv3-1.2.0.jar:?]
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:145) ~[org.eclipse.paho.client.mqttv3-1.2.0.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_162]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_162]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_162]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_162]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_162]
2018-08-27 10:47:04,488 ERROR --- MqttFX ClientModel             : Please verify your Settings (e.g. Broker Address, Broker Port & Client ID) and the user credentials!
org.eclipse.paho.client.mqttv3.MqttException: Broker unavailable
	at org.eclipse.paho.client.mqttv3.internal.ExceptionHelper.createMqttException(ExceptionHelper.java:31) ~[org.eclipse.paho.client.mqttv3-1.2.0.jar:?]
	at org.eclipse.paho.client.mqttv3.internal.ClientState.notifyReceivedAck(ClientState.java:988) ~[org.eclipse.paho.client.mqttv3-1.2.0.jar:?]
	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:145) ~[org.eclipse.paho.client.mqttv3-1.2.0.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[?:1.8.0_162]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_162]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_162]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_162]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_162]
2018-08-27 10:47:04,492  INFO --- ScriptsController              : Clear console.
2018-08-27 10:47:04,495 ERROR --- BrokerConnectService           : Broker unavailable
```

Could you please guide me to the right direction to access the server so that I apply the configuration to Kura to connect to Hono? Thank you very much for your consideration.

**Comment by MMaiero** [comment]
If the error is "broker unavailable" it has no connection to the password or credential used.
The broker seems unreachable. If you try to access to the server with the command 

> telnet hono.eclipse.org 1884

What do you get?

In any case, please refer to the Hono team that can give you more insight on their product.

Best regards,
Matteo

**Comment by mohabh88** [comment]
Thank you Matteo. Telnet disconnects right after a couple of seconds

`telnet hono.eclipse.org 1884`

```
Trying 35.185.31.194...
Connected to hono.eclipse.org.
Escape character is '^]'.
Connection closed by foreign host.
```

**Comment by mohabh88** [comment]
Thank you so much Matteo for your efforts. Now I am able to connect Kura to Hono's Sandbox :)

**Comment by MMaiero** [comment]
Thank you for the update.

Closing for now.
